File: bindings/python/examples/onion_service.py
"""Hosting an onion service.

python examples/onion_service.py
"""

import hypertor

# The nickname is the identity: the .onion address is derived from a key filed
# under it, and arti's default state directory is persistent. Relaunching with
# the same nickname republishes the *same* address, with or without state_dir —
# that argument only chooses where the key lives.
app = hypertor.OnionApp("my-service", state_dir="./onion-state")

_USERS = {"1": {"id": "1", "name": "Alice"}}


@app.get("/")
def home(request):
    return "<h1>Hello from a .onion service</h1>"


@app.get("/health")
def health(request):
    return {"status": "ok"}


@app.get("/users/{user_id}")
def get_user(request):
    user = _USERS.get(request.params["user_id"])
    if user is None:
        # Flask's convention: (body, status) and (body, status, headers).
        return {"error": "no such user"}, 404
    return user


@app.post("/users")
def create_user(request):
    user = request.json()
    _USERS[user["id"]] = user
    return user, 201, {"location": f"/users/{user['id']}"}


@app.post("/echo")
def echo(request):
    return {"received": request.json()}


if __name__ == "__main__":
    # Prints the .onion address, then serves until Ctrl-C.
    app.run()

File: examples/onion_service.rs
//! Hosting an onion service.
//!
//! ```console
//! $ cargo run --example onion_service --features server
//! ```
//!
//! Proof-of-work needs the `pow` feature, which pulls in LGPL-3.0 dependencies
//! and is therefore opt-in:
//!
//! ```console
//! $ cargo run --example onion_service --features "server,pow"
//! ```

use hypertor::{OnionApp, OnionService, ServeResponse};

#[tokio::main]
async fn main() -> hypertor::Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter("hypertor=info")
        .init();

    let app = OnionApp::new()
        .get("/", |_req| async {
            ServeResponse::html("<h1>Hello from a .onion service</h1>")
        })
        .get("/health", |_req| async {
            ServeResponse::json(&serde_json::json!({"status": "ok"}))
        })
        .get("/users/{id}", |req| async move {
            let id = req.param("id").unwrap_or("unknown").to_string();
            ServeResponse::json(&serde_json::json!({"id": id}))
        })
        .post("/echo", |req| async move {
            match req.text() {
                Ok(body) => ServeResponse::json(&serde_json::json!({"received": body})),
                Err(e) => {
                    ServeResponse::new(http::StatusCode::BAD_REQUEST).with_body(e.to_string())
                }
            }
        });

    // The nickname is the identity: the address is derived from a key arti
    // files under it. Relaunching with the same nickname republishes the same
    // .onion address — including without `state_dir`, which only chooses *where*
    // that key lives rather than whether it is kept.
    let service = OnionService::builder()
        .nickname("hypertor-example")?
        .state_dir("./onion-state")
        // Proof-of-work only engages under load, so it costs ordinary clients
        // nothing while raising the price of an introduction flood. Requires
        // the `pow` feature.
        .proof_of_work(cfg!(feature = "pow"))
        .launch()
        .await?;

    let serving = app.serve_on(service).await?;

    println!("\n  🧅 {}\n", serving.onion_address());
    println!("  reach it with any Tor-capable client, for example:");
    println!(
        "    curl --socks5-hostname 127.0.0.1:9050 http://{}/",
        serving.onion_address()
    );
    println!("  (run `cargo run --example socks_proxy` for that proxy)\n");

    serving.wait().await
}
